lib/client/alpn.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	alpn "github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// ALPNAuthTunnel contains the required methods to create a local ALPN proxy.
+type ALPNAuthTunnel interface {
```

Comment:
I think that interface name should be based what interfaces does and what functionality it provides  instead of naming it based on high level usage. 

The `GetClusterCACert`, `GetCurrentUser` and `GenerateUserCerts` methods are subsets of gRPC teleport auth client. What would you say about just naming it as `authClient` ?

Reply:
Sorry, I'm not sure I understood
Are you suggesting the rename of the interface from `ALPNAuthTunnel` to `authClient`?
`authClient` feels too generic for a `client` package.
We are also re-using this interface in `conntest.ClientDatabaseConnectionTester`, so moving this to a private interface wouldn't work without further changes.

I agree with changing the name to something that better fits what it defines.

Maybe `ALPNAuthClient`?

We could move this to a sub-package and use `authClient`, but it seems overkill.
I would  follow a pattern here if there's one.

Reply:
The `ALPNAuthClient` sound better but wonder if we can narrow the scope of this interfaces. 
Why the `ALPNAuthTunnel` needs to be defined in `lib/client` if it is only used by `lib/client/conntest/` Can we just implement alpn logic under `lib/client/conntest/` ?

Reply:
I can move it into `conntest` but the goal was to create a generic `ALPNAuthTunnel` that could be used not only by `conntest` but also by `tsh`.

@greedy52 was also questioning if we could use this `ALPNAuthTunnel` in tsh.